proposals/span-safety.md

```diff
@@ -0,0 +1,104 @@
+
+Example:
+
+```cs
+
+Span<T> Caller()
+{
+	Span<T> safeToReturn = GetSafeToReturnSpan(array);
+
+	if (TodaIsNotFriday())
+	{
+		// callee can "poison" any assumption that we have made about safeToReturn
+		Callee(ref safeToReturn);
+	}
+
+	// is this safe?
+	return safeToReturn;
+}
+
+void Callee(ref Span<int> arg1)
+{
+	// local references are escaping. 
+	// what rule would disallow this?
+	arg1 = new Span<T>(localloc(...));
+}
+
+```
+
+Note: returning by a ordinary writeable reference appears to be ok by itself, but with values on heap being impossible, values in local frame being not returnable and variables passed from the caller being readonly, it may not matter.
+
+Another way to look at this is that writeable references to a ref-like variable would never be exposed outside of the containing method, except, of course, if that method is a constructor which has to take `this` by a writeable ref.
+
+## `ref-like` types must be readonly structs. ##
+
+- ref-like types must be readonly. 
+
+This basically follows from the previous "only pass by read-only references" rule when applied to "this" inside members of ref-like types.
+
+## not all `ref-like` variables are safe to return. ##
+
+In addition to the regular rules that govern ref-returnability, there are additional rules that apply to by-value returns of ref-like typed variables.
+
+* ref-like variables, free of local references are safe to return.
+* ref-like parameters are safe to return.
+* ref-like fields are safe to return as long as the receiver is safe to return
+* a ref-like value, returned from an expression (unary, binary, conditional...) is safe to return if all ref and ref-like operands were safe to return.
+* “this” is not considered safe to return inside the members of ref-like types
+* a ref-like value, returned from a method is safe to return if all ref/out variables and ref-like values passed to that method as formal parameters were safe to return.
+Specifically it is irrelevant if receiver is safe to return, regardless whether receiver is a struct, class or typed as a generic type parameter.
+
+## Reassignment of `ref-like` variables. ##
+
+Reassignment of ref-like variables to new values can easily circumvent scoping/life-time invariants. A particular concern is when a ref-like variable with a bigger span is assigned a ref-like variable that refers to a variable from an inner scope.
+It would be even worse if the LHS is a returnable variable, but the RHS is not a returnable value.  
+
+There are ways to prevent such problems:
+
+The most permissive approach would be alias/value tracking via fixed point data-flow analysis. It is very expensive and complex approach to be practical, but will be called out here for completeness.
+
+Simpler solutions are:
+- make ref-like variables "single-assignment" similarly to the solution we have for ordinary ref variables.
```

Comment:
@KrzysztofCwalina - would it be ok to allow this only for spans that are free from referring to local data?

With local data you would have to do  'var anotherSpan = span.Slice(count);` 

The reason is that inside a method could be many nested scopes and it would be very hard to enforce that variables from outer scopes do not end up referring indirectly to variables from inner scopes if they can be reassigned.  
On the other hand, everything outside of the method lives as long as you have a reference to it, so it is safe to assign to anything.